src/algebra/algebra/spectrum.lean

```diff
@@ -0,0 +1,187 @@
+
+universes u v
+
+namespace spectrum
+
+section defs
+
+variables (R : Type u) {A : Type v}
+variables [comm_ring R] [ring A] [algebra R A]
+
+-- definition and basic properties
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, the *resolvent* of `a : A`
+is the `set R` consisting of those `r : R` for which `r•1 - a` is a unit of the
+algebra `A`.  -/
+definition resolvent (a : A) : set R :=
+{ r : R | is_unit (r • 1 - a) }
+
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, the *spectrum* of `a : A`
+is the `set R` consisting of those `r : R` for which `r•1 - a` is not a unit of the
+algebra `A`.
+
+The spectrum is simply the complement of the resolvent.  -/
+definition spectrum (a : A) : set R :=
+(resolvent R a)ᶜ
+
+end defs
+
+variables {R : Type u} {A : Type v}
+variables [comm_ring R] [ring A] [algebra R A]
+
+local notation `σ` := spectrum R
+
+lemma mem_spectrum_iff_not_unit {r : R} {a : A} :
+  r ∈ σ a ↔ ¬ is_unit (r • 1 - a) :=
+by { tidy }
+
+lemma not_mem_spectrum_iff_unit {r : R} {a : A} :
+  r ∉ σ a ↔ is_unit (r • 1 - a) :=
+by { apply not_iff_not.mp, simp [set.not_not_mem, mem_spectrum_iff_not_unit] }
+
+lemma mem_resolvent_of_left_right_inverse {r : R} {a b c : A}
+  (h₁ : (r • 1 - a) * b = 1) (h₂ : c * (r • 1 - a) = 1) :
+  r ∈ resolvent R a :=
+units.is_unit ⟨r•1 - a, b, h₁, by rwa ←left_inv_eq_right_inv h₂ h₁⟩
+
+lemma mem_resolvent_iff {r : R} {a : A} :
+  r ∈ resolvent R a ↔ is_unit (r•1 - a) :=
+by { tidy }
+
+-- products of scalar units and algebra units
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, and units `r : units R`
+and `a : units A`, then `unit_mul_unit r a` constructs a `units A` with value `r•a`. -/
+
+lemma is_unit.smul_smul_sub_smul_iff {r : units R} {s : R} {a : A} :
+  is_unit (r • s • 1 - r • a) ↔ is_unit (s • 1 - a) :=
+begin
+  split,
+  { intro h',
+    have inv_smul_eq : r⁻¹•(r•s•1 - r•a) = s•1 - a,
+      by simp [smul_sub, smul_smul],
+    rw ←inv_smul_eq,
+    exact (r⁻¹ • h'.unit).is_unit, },
+  { intro h',
+    rw ←smul_sub,
+    exact (r • h'.unit).is_unit, },
+end
+
+lemma is_unit.smul_smul_sub_iff_is_unit_smul_sub_smul {r : units R} {s : R} {a : A} :
+  is_unit (r • s • 1 - a) ↔ is_unit (s • 1 - r⁻¹ • a) :=
+by { have h_eq : r•s•1 - r•(r⁻¹•a) = r•s•1 - a, by simp,
+     rw [←h_eq,is_unit.smul_smul_sub_smul_iff], }
+
+lemma is_unit.smul_sub_iff_is_unit_sub_smul {r : units R} {a : A} :
+  is_unit (r • 1 - a) ↔ is_unit (1 - r⁻¹ • a) :=
+begin
+  have with_smul_one : is_unit (r•(1 : R)•1 - a) ↔ is_unit ((1 : R)•1 - r⁻¹•a),
+    by exact is_unit.smul_smul_sub_iff_is_unit_smul_sub_smul,
+  simpa using with_smul_one,
+end
+
+lemma add_mem_spectrum {a : A} {r s : R} :
+  r ∈ σ a ↔ r + s ∈ σ (s • 1 + a) :=
+begin
+  apply not_iff_not.mpr,
+  simp only [mem_resolvent_iff],
+  have h_eq : (r+s)•1 - (s•1 + a) = r•1 - a,
+    by { rw add_smul, noncomm_ring },
+  simp [h_eq],
+end
+
+lemma mul_mem_spectrum_mul {a : A} {s : R} {r : units R} :
+  r • s ∈ σ (r • a) ↔ s ∈ σ a :=
```

Comment:
Is the following true? (with `open_locale pointwise`)
```lean
   σ (r • a) = r • σ a :=
```

Reply:
Yes, this is nice. I didn't realize we had access to such nice ways of stating this.